protocol/eth-shared-lockbox.md

```diff
@@ -0,0 +1,164 @@
+// Current dependency set list
+uint256[] public dependencySet;
+
+// Function to add a chain to the dependency set
+function addChain(uint256 chainId) external onlyOwner {
+		require(systemConfigInterops[chainId] != address(0), "DependencySetManager: Chain not registered");
+		require(!isRegistered[chainId], "DependencySetManager: Chain already in dependency set");
+
+		// For the chain being added, call addDependencies with existing chainIds
+		if (dependencySet.length > 0) {
+				for (uint256 i = 0; i < dependencySet.length; i++) {
+            ISystemConfigInterop(systemConfigInterops[chainId]).addDependency(dependencySet[i]);
+            }
+		}
+
+		// For each existing chain, call addDependencies with the new chainId
+		for (uint256 i = 0; i < dependencySet.length; i++) {
+				ISystemConfigInterop(systemConfigInterops[dependencySet[i]]).addDependency(chainId);
+    }
+    // Update the dependency set
+    dependencySet.push(chainId);
+    isRegistered[chainId] = true;
+
+    emit ChainAdded(chainId);
+}
+    
+```
+
+For `removeChain`, the process would follows the same logic.
+
+Note that, under the specified flow, the dependency set consistently maintains the form of a [complete graph](https://en.wikipedia.org/wiki/Complete_graph) at all times.
+
+### `addDependencies` and `removeDependencies` in `SystemConfigInterop` and `L1BlockInterop`
+
+The purpose of these functions is to extend the current `SystemConfigInterop` implementation by allowing multiple `chainId` values to be added in a single call by passing them as an array. There are a few ways to pass these values through `setConfig`: one is making a single call and triggering the `TransactionDeposited` event, or by looping through the current `addDependency`/`removeDependency` functions.
+
+With this optimization, the `DependencySetManager` can simplify the number of calls, as follows:
+
+```solidity
+function addChain(uint256 chainId) external onlyOwner {
+  // ...
+	if (dependencySet.length > 0) {
+				// If we can input an array chainId values thorugh addDependencies function
+				ISystemConfigInterop(newChainSystemConfig).addDependencies(dependencySet);
+	}		
+	// ...
+```
+
+### `SharedLockbox` implementation
+
+A minimal set of functions should include:
+
+- `lockETH`: Accepts ETH from the `depositTransaction` originating from a valid `OptimismPortal2`.
+- `unlockETH`: Releases ETH from the `finalizeWithdrawalTransaction` originating from a valid `OptimismPortal2`.
+
+Access control for `lockETH` and `unlockETH` is validated against the list reated by `registerChain`.
+
+### `OptimismPortal2` upgrade process
+
+A one-time L1 liquidity migration is required for each approved chain. By using an intermediate contract during the upgrade, all ETH held by the `OptimismPortal` can be transferred to the `SharedLockbox` within the `initialize` function. After this migration, the `OptimismPortal` is upgraded to the new version with the desired functionality.
```

Comment:
What alternative migration flows have we considered and what are the tradeoffs? For example, we could just add a `forwardETH()` method to the portals that forwards all ETH balance to the lockbox whenever it's called, and call it from the `initialize` function

Reply:
We are exploring various migration options, with the goal of minimizing the amount of permanently added code. We’ll share some insights on this soon.